stl/inc/xstring

```diff
@@ -2642,6 +2642,44 @@ private:
     }
 
 public:
+#if _HAS_CXX20
+    void _Move_assign_from_buffer(_Elem* const _Right, const size_type _Size, const size_type _Res) {
+        // Move assign from a buffer, used by basic_stringbuf
+        _Tidy_deallocate();
+        _Right[_Size]  = _Elem();
+        auto& _My_data = _Mypair._Myval2;
+        _Construct_in_place(_My_data._Bx._Ptr, pointer_traits<pointer>::pointer_to(*_Right));
+        _My_data._Mysize = _Size;
+        _My_data._Myres  = _Res;
+    }
+
+    struct _Released_buffer {
+        _Elem* _Ptr;
+        size_type _Size;
+        size_type _Res;
+    };
+
+    _NODISCARD _Released_buffer _Release_to_buffer() {
+        // Release to a buffer, or allocate a new one if in small string mode
+        _Released_buffer _Result;
+        auto& _My_data = _Mypair._Myval2;
+        if (_My_data._Large_string_engaged()) {
+            _Result._Ptr      = _Unfancy(_My_data._Bx._Ptr);
+            _My_data._Bx._Ptr = nullptr;
+            _Result._Size     = _STD exchange(_My_data._Mysize, size_type{0});
+            _Result._Res      = _STD exchange(_My_data._Myres, size_type{0});
+            _Tidy_init();
+        } else {
+            _Result._Ptr = _Unfancy(_Getal().allocate(_BUF_SIZE));

```

Comment:
I think this is a bug: our invariant is that _Myres == _BUF_SIZE means the string is in small mode, but if you _Move_assign_from_buffer this back in the resulting string will try to set _Ptr  but be in small mode. Maybe choose 2*_BUF_SIZE or similar?

Probably deserves a regression test added for this corner case (move from small string to stringbuf back to small string; or maybe the full cross product to/from small/big)

Reply:
It seems that small or big mode is distinguished by reserved size. I'll try to make a small string if possible.